### ISSUE TITLE
Configure Reload MP Config cache through System Property or Env Variable: (was : Reload of ConfigSource @ConfigProperty Provider<String> is somehow random) /FISH-427

### ISSUE THREAD

**Issue by sterlp** [payara/Payara#4839]
# Description #
----------

Based on the documentation the **ConfigSource** it can decide if it is dynamic or static. The problem with a "dynamic" **ConfigSource** is, that the payara micro seems somehow to cache the values. Neither the scope of the hosting class not the usage of **Provider** or **Config** allows to get reliable results, means get the refreshed value.
@ConfigProperty Provider<String>

https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#custom-configsources-via-configsourceprovider

## Expected Outcome

Usage of **@ConfigProperty Provider<>** will resolve the value from the underlinging **ConfigSource**.

## Current Outcome

The reload will happen randomly somehow // not at once, as if the container caches the value somehow.

## Steps to reproduce (Only for bug reports) 

1. Clone https://github.com/sterlp/training/tree/master/spring-cloud-config
2. Start Spring Config Server
3. Start payara micro: **mvn payara-micro:start**
4. Request GET http://localhost:8001
5. Check Logoutput **-> getValue: ** appears randomly

## Environment ##

- **Payara Version**: 5.2020.3
- **Edition**: Micro
- **JDK Version**:  11
- **Operating System**: Mac

**Comment by smillidge** [comment]
There is a 60s cache of the values see https://docs.payara.fish/community/docs/5.2020.4/documentation/microprofile/config.html it can be changed with a preboot or postboot command.

**Comment by sterlp** [comment]
Thanks for pointing this out, now it makes sense. 

Is where a way to set it as a configuration parameter either in micro profile configuration file or as env. parameter? The documentation points to an asadmin command. I would prefere using payara micro config or a java env parameter. Tried to google it, but had no real luck finding the parameter.

I found only this documentation sending a command using the "api"
https://docs.payara.fish/enterprise/docs/5.21.0/documentation/payara-micro/asadmin/using-the-payara-micro-api.html

But this somehow conflicts with the "payara-micro-maven-plugin" -- as this plugin wraps the startup of the container -- means I don't really have access to it -- despite using a static reference maybe which is somehow "odd". It would mean I cannot use the plugin but have to add a main method.

**Comment by sterlp** [comment]
Thanks a lot guys for the support realy appreciate it! Maybe it would make sense to and the info to the doc.

**Comment by sterlp** [comment]
@jbee I tried several ways to use it, seems it has no effect:

1. Java Argument
`mvn payara-micro:start -Dmp.config.cache.duration=0`

2. Microprofile config
![image](https://user-images.githubusercontent.com/774984/91668049-8b265680-eb09-11ea-9d7d-d37cf8e1e7d0.png)


3. Application props:
```java
@ApplicationPath("/")
public class ApplicationConfig extends Application {
    @Override
    public Map<String, Object> getProperties() {
        Map<String, Object> props = new HashMap<>();
        props.put("jersey.config.server.disableMoxyJson", true);
        props.put("mp.config.cache.duration", "0");
        return props;
    }
}
```
did I miss  something?

**Comment by sterlp** [comment]
I looked a bit into the code and it is a bit odd:

The parameter is needed to build the "server config:"
![image](https://user-images.githubusercontent.com/774984/91668194-20761a80-eb0b-11ea-86e4-39316786aa40.png)

But get the config we the "server config" is used:
![image](https://user-images.githubusercontent.com/774984/91668204-3f74ac80-eb0b-11ea-8675-e1f312afc73e.png)
Based on the flow this is always "null", right?

I am not sure if you currently can set it from outside ...

**Comment by jbee** [comment]
@sterlp I have a look next week, you might be right that the order makes the MP config ineffective. In that case there is only the asadmin command you can use to configure the cache.

**Comment by jbee** [comment]
I guess we had not thought about Payara Micro and that it apparently does not have an `ApplicationInfo` which means it uses the server level `Config`. At that point we are unable to read the cache duration from the config we are about to bootstrap so the MP configuration becomes ineffective. 

I talked to  @MattGill98 and we agreed to feed the ticket into the community process so the hen-egg issue can be addressed.
This is likely going to happen in connection to update to MP config standard that is ahead of us.

**Comment by rdebusscher** [comment]
Hi Paul,

I have created the internal ticket `FISH-427` to implement this improvement in the future.

Regards
Rudy

**Comment by MattGill98** [comment]
This looks like what I encountered in the Metrics upgrade. Perhaps https://github.com/payara/Payara/pull/5002/files/cd18a2e0a49f2aed17959eb32aff96e140e78536..7cc892da1ede4aee369527829c96bf12ce4a3660 fixes it?

**Comment by AlanRoth** [comment]
Hi @sterlp, starting today we are introducing some relevant changes to our Github Issues Community Policy, which you can read [here](https://github.com/payara/Payara/blob/master/SUPPORT.adoc). The most significant change is that we want to introduce open votes for all suggested enhancements and improvements to the Payara Platform, like the one described in this issue. To this effect we are going to transition this issue to Voting status and gauge the community interest on your proposal. We know that previously we accepted your requested and escalated it to the our Platform Development team, but after careful deliberation and analysis we want to improve our community engagement in a way that’s productive and relevant for our overall user community. We hope that you understand this decision.

**Comment by AlanRoth** [comment]
Hi @sterlp ,

`FISH-427` has been resolved in release [5.2021.1](https://docs.payara.fish/community/docs/Release%20Notes/Release%20Notes%205.2021.1.html) with this PR to Upgrade MP Metrics to 3.0; https://github.com/payara/Payara/pull/5002

I will be closing this issue.

Thank you,
Alan
